packages/gatsby-script/src/gatsby-script.tsx

```diff
@@ -0,0 +1,46 @@
+import React, { useEffect } from "react"
+import { ReactElement } from "react"
+
+export enum ScriptStrategy {
+  preHydrate = `pre-hydrate`,
+  postHydrate = `post-hydrate`,
+  idle = `idle`,
+}
+
+// eslint-disable-next-line @typescript-eslint/naming-convention
+export interface ScriptProps {
+  src: string
+  strategy?: ScriptStrategy
+}
+
+export function Script(props: ScriptProps): ReactElement {
+  const { src, strategy = ScriptStrategy.postHydrate } = props || {}
+
+  useEffect(() => {
+    switch (strategy) {
+      case ScriptStrategy.postHydrate:
+        injectScript(src, strategy)
+        break
+      case ScriptStrategy.idle:
+        requestIdleCallback(() => {
+          injectScript(src, strategy)
+        })
+        break
+      default:
+        return
+    }
+  }, [])
+
+  if (strategy === ScriptStrategy.preHydrate) {
+    return <script async src={src} data-strategy={strategy} />
```

Comment:
Question: do we want to keep the preHydrate <script> element in same "DOM location" as would be indicated by react tree (current implementation) or would we want to just collect them and render at the bottom (so like "postBodyComponents" in our SSR template ( https://github.com/gatsbyjs/gatsby/blob/f63d13db0c635c0ade66169982e061630306fca0/packages/gatsby/cache-dir/default-html.js#L23 )

With work in https://github.com/gatsbyjs/gatsby/pull/35408 we are looking to get critical assets (non-js) to load earlier than scripts and just thinking that putting scripts potentially in the middle of DOM might interact weirdly with those changes?

Reply:
Those are good points, it would also be one less thing to explain in the documentation if all the strategies (except for `experimental-partytown`) behave the same when mutating the DOM

Reply:
Thought more on this and we'd have to parse and manipulate the DOM during SSR (or expose `document` somehow), so will keep the behavior as is for now. Can revisit the idea in the future